File: fplbr-frontend/src/types/PlanoDeVooDTO.ts
export type ModoPlano = 'PVC' | 'PVS'

export interface BotesInfo {
  possui: boolean
  numero?: number
  capacidade?: number
  c?: boolean
  cor?: string
}

export interface InformacaoSuplementar {
  autonomia: string // HHmm
  pob: number // Pessoas a bordo (máximo 3 dígitos)
  radioEmergencia?: string[] // U, V, E
  sobrevivencia?: string[] // S, P, D, M, J
  coletes?: string[] // se incluir J, habilita demais opções L,F,U,V
  botes?: BotesInfo
  corEMarcaAeronave: string
  observacoes?: string
  pilotoEmComando: string
  anacPrimeiroPiloto: string
  anacSegundoPiloto?: string
  telefone: string // 10-11 dígitos
  n?: boolean // se true, observações obrigatório
}

export interface OutrasInformacoes {
  sts?: string // Status especial
  pbn?: string // Performance Based Navigation
  nav?: string // Navegação
  com?: string // Comunicação
  dat?: string // Dados
  sur?: string // Vigilância
  dep?: string // Aeródromo de partida
  dest?: string // Aeródromo de destino
  reg?: string // Registro da aeronave
  eet?: string // ex: "SBC0005 SBBS0150"
  sel?: string // SELCAL
  typ?: string // Tipo de aeronave
  code?: string // Código
  dle?: string // Delay
  opr: string // Operador
  orgn?: string // Origem
  per?: string[] // PVC: A, B, C, D, E, H
  altn?: string // Alternativo
  ralt?: string // Rota alternativa
  talt?: string // Takeoff alternativo
  rif?: string // Route if
  rmk?: string // Observações
  from: string // ICAO 4 letras
  dof: string // ddmmaa
}

export interface PlanoDeVooDTO {
  identificacaoDaAeronave: string
  indicativoDeChamada: boolean
  regraDeVooEnum: 'I' | 'V' | 'Y' | 'Z'
  tipoDeVooEnum: 'G' | 'S' | 'N' | 'M' | 'X'
  numeroDeAeronaves: number
  tipoDeAeronave: string
  categoriaEsteiraTurbulenciaEnum: 'L' | 'M' | 'H' | 'J'
  equipamentoCapacidadeDaAeronave: string[]
  vigilancia: string[]
  aerodromoDePartida: string
  horaPartida: string // ISO UTC
  aerodromoDeDestino: string
  tempoDeVooPrevisto: string // HHmm
  aerodromoDeAlternativa: string
  aerodromoDeAlternativaSegundo?: string
  velocidadeDeCruzeiro: string
  nivelDeVoo: string // VFR | Axxx | Fxxx
  rota: string
  outrasInformacoes: OutrasInformacoes
  informacaoSuplementar: InformacaoSuplementar
  modo: ModoPlano
}

File: backend/src/main/java/br/com/fplbr/pilot/infrastructure/web/dto/PlanoDeVooDTO.java
package br.com.fplbr.pilot.infrastructure.web.dto;

import jakarta.validation.Valid;
import jakarta.validation.constraints.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.ToString;

import java.time.OffsetDateTime;
import java.util.List;

@Getter
@Setter
@ToString
@Builder
@AllArgsConstructor
@NoArgsConstructor
public class PlanoDeVooDTO {

    @NotBlank
    @Pattern(regexp = "^[A-Z0-9]{3,7}$")
    private String identificacaoDaAeronave;

    @NotNull
    private Boolean indicativoDeChamada;

    @NotBlank
    @Pattern(regexp = "^[IVYZ]$")
    private String regraDeVooEnum;

    @NotBlank
    @Pattern(regexp = "^[GSNMX]$")
    private String tipoDeVooEnum;

    @NotNull
    @Min(1)
    private Integer numeroDeAeronaves;

    @NotBlank
    @Pattern(regexp = "^[A-Z0-9]{2,4}$")
    private String tipoDeAeronave;

    @NotBlank
    @Pattern(regexp = "^[LMHJ]$")
    private String categoriaEsteiraTurbulenciaEnum;

    @NotNull
    @Size(min = 1)
    private List<@NotBlank String> equipamentoCapacidadeDaAeronave;

    @NotNull
    @Size(min = 1)
    private List<@NotBlank String> vigilancia;

    @NotBlank
    @Pattern(regexp = "^[A-Z]{4}$")
    private String aerodromoDePartida;

    @NotNull
    private OffsetDateTime horaPartida; // ISO UTC esperado

    @NotBlank
    @Pattern(regexp = "^[A-Z]{4}$")
    private String aerodromoDeDestino;

    @NotBlank
    @Pattern(regexp = "^([0-1]\\d|2[0-3])[0-5]\\d$")
    private String tempoDeVooPrevisto; // HHmm

    @NotBlank
    @Pattern(regexp = "^[A-Z]{4}$")
    private String aerodromoDeAlternativa;

    @Pattern(regexp = "^$|^[A-Z]{4}$")
    private String aerodromoDeAlternativaSegundo; // opcional

    @NotBlank
    @Pattern(regexp = "^[NKM][0-9]{4}$")
    private String velocidadeDeCruzeiro;

    @NotBlank
    @Pattern(regexp = "^(VFR|[AF][0-9]{3})$")
    private String nivelDeVoo;

    @NotBlank
    private String rota;

    @Valid
    @NotNull
    private OutrasInformacoes outrasInformacoes;

    @Valid
    @NotNull
    private InformacaoSuplementar informacaoSuplementar;

    @NotBlank
    @Pattern(regexp = "^(PVC|PVS)$")
    private String modo; // PVC | PVS

    // ==== Tipos aninhados ====

    @Getter
    @Setter
    @ToString
    @Builder
    @AllArgsConstructor
    @NoArgsConstructor
    public static class OutrasInformacoes {
        @Pattern(regexp = "^$|^(?:[A-Z]{3,4}[0-9]{4})(?:\\s+[A-Z]{3,4}[0-9]{4})*$")
        private String eet; // SBC0005 SBBS0150 ...

        @NotBlank
        private String opr;

        @NotBlank
        @Pattern(regexp = "^[A-Z]{4}$")
        private String from;

        private List<@Pattern(regexp = "^[A-Z]$") String> per; // PVC

        private String rmk; // deve conter REA quando rota tiver REA

        @NotBlank
        @Pattern(regexp = "^\\d{6}$")
        private String dof; // ddmmaa
    }

    @Getter
    @Setter
    @ToString
    @Builder
    @AllArgsConstructor
    @NoArgsConstructor
    public static class BotesInfo {
        @NotNull
        private Boolean possui;
        @Min(1)
        private Integer numero;
        @Min(1)
        private Integer capacidade;
        private Boolean c; // abrigo
        private String cor;
    }

    @Getter
    @Setter
    @ToString
    @Builder
    @AllArgsConstructor
    @NoArgsConstructor
    public static class InformacaoSuplementar {
        @NotBlank
        @Pattern(regexp = "^([0-1]\\d|2[0-3])[0-5]\\d$")
        private String autonomia; // HHmm

        private List<@Pattern(regexp = "^[UVE]$") String> radioEmergencia;
        private List<@Pattern(regexp = "^[SPDMJ]$") String> sobrevivencia;
        private List<@Pattern(regexp = "^[JLFUV]$") String> coletes;

        @Valid
        private BotesInfo botes;

        @NotBlank
        private String corEMarcaAeronave;

        private String observacoes;

        @NotBlank
        private String pilotoEmComando;

        @NotBlank
        private String anacPrimeiroPiloto;

        private String anacSegundoPiloto;

        @NotBlank
        @Pattern(regexp = "^\\d{10,11}$")
        private String telefone;

        private Boolean n; // se true, observações obrigatório
    }
}


